lib/lE_mirage.ml

```diff
@@ -0,0 +1,159 @@
+    {
+      Alpn.alpn = alpn_of_tls_connection;
+      Alpn.peer = peer_of_tls_connection;
+      Alpn.injection;
+    }
+
+  let with_lets_encrypt_certificates ?(port = 443) stackv4v6 ~production config
+      handler =
+    let certificates = ref None in
+    let stop_http_server = Lwt_switch.create () in
+    let stop_alpn_server = Lwt_switch.create () in
+    let mutex = Lwt_mutex.create () in
+
+    let ctx =
+      let gethostbyname dns domain_name =
+        DNS.gethostbyname dns domain_name >>? fun ipv4 ->
+        Lwt.return_ok (Ipaddr.V4 ipv4) in
+      LE.ctx ~gethostbyname
+        ~authenticator:(Result.get_ok (Nss.authenticator ()))
+        (DNS.create stackv4v6) stackv4v6 in
+
+    let rec fill_certificates () =
+      LE.provision_certificate ~production config ctx >>= function
+      | Error _ as err ->
+          Lwt_switch.turn_off stop_http_server >>= fun () ->
+          Lwt_switch.turn_off stop_alpn_server >>= fun () -> Lwt.return err
+      | Ok v ->
+          Lwt_mutex.with_lock mutex (fun () ->
+              certificates := Some v ;
+              Lwt.return_unit)
+          >>= fun () ->
+          (* TODO(dinosaure): should we [reneg] all previous connections? *)
+          Time.sleep_ns (Duration.of_day 80) >>= fill_certificates in
+
+    let handshake tcp =
+      Lwt_mutex.with_lock mutex (fun () -> Lwt.return !certificates)
+      >>= function
+      | None -> Lwt.return_error `No_certificates
+      | Some certificates -> (
+          let cfg =
+            Tls.Config.server ~alpn_protocols:[ "http/1.1"; "h2" ] ~certificates
+              () in
+          Paf.TLS.server_of_flow cfg tcp >>= function
+          | Ok flow -> Lwt.return_ok (Paf.TCP.dst tcp, flow)
+          | Error `Closed -> Lwt.return_error (`Write `Closed)
+          | Error err ->
+              let err = Rresult.R.msgf "%a" Paf.TLS.pp_write_error err in
+              Paf.TCP.close tcp >>= fun () -> Lwt.return_error err) in
+    let module R = (val Mimic.repr Paf.tls_protocol) in
+    let request flow edn reqd protocol =
+      match flow with
+      | R.T flow -> handler.Alpn.request flow edn reqd protocol
+      | _ -> assert false in
+
+    let alpn_service =
+      Alpn.service info { handler with request } handshake Paf.accept Paf.close
+    in
+    let http_service =
+      let request_handler _ edn reqd =
+        let request = Httpaf.Reqd.request reqd in
+        match String.split_on_char '/' request.Httpaf.Request.target with
+        | [ ""; _p1; _p2; _token ] -> LE.request_handler edn reqd
+        | _ -> redirect config port reqd in
+      let error_handler _dst ?request err _ =
+        Logs.err (fun m ->
+            m "error %a while processing request %a" pp_error err
+              Fmt.(option ~none:(any "unknown") Httpaf.Request.pp_hum)
+              request) in
+      Paf.http_service ~error_handler request_handler in
+
+    Paf.init ~port:80 (Stack.tcp stackv4v6) >>= fun http ->
```

Comment:
Is this still necessary? I was under the impression that the Alpn protocol would remove the need for a hardcoded HTTP service on port 80.

Reply:
The HTTP/80 webserver is mainly here to solve 2 issues:
- redirect any request from `http://hostname/` to `https://hostname/`
- still be able to launch the let's encrypt challenge when the certificate expired

Reply:
I guess that's where my confusion comes from, eventually @dinosaure as well.

With RFC 8737, let's encrypt allows a ALPN challenge which works as follows:
- the ACME server (from let's encrypt) connects via http + tls (client) to the domain name of the certificate
- in this connection, the ALPN `tls-alpn-01` is used
- our server is supposed to use a specific certificate (self-signed with some ACME extension) and private key to show that our let's encrypt client is in control of the web server

Now, with the current setup of TLS and certificate selection, I don't think we can easily get this up.

So, the current approach, as implemented by @dinosaure, looks fine to me. Future work may include to extend the TLS config in respect of certificate selection (though the logic is already pretty complex, unfortunately).

TL;DR: I opened https://github.com/mirleft/ocaml-tls/issues/465 to track the above, and once that is achieved, we can come back to paf.